project_euler/problem_092/sol1.py

```diff
@@ -0,0 +1,141 @@
+"""
+Project Euler Problem 92:https://projecteuler.net/problem=92
+
+A number chain is created by continuously adding the square of the
+digits in a number to form a new number until it has been seen before.
+
+For example,
+44 → 32 → 13 → 10 → 1 → 1
+85 → 89 → 145 → 42 → 20 → 4 → 16 → 37 → 58 → 89
+
+Therefore any chain that arrives at 1 or 89 will
+become stuck in an endless loop.
+What is most amazing is that EVERY starting number
+will eventually arrive at 1 or 89.
+
+How many starting numbers below ten million will arrive at 89?
+
+For a n digit number abc..p(n digits) if it arrive at 89
+then any permutation of digits like bad..p will also arrive at 89
+and same in the case if it does not.
+So if I check for a single permutation whether it arrives or not.
+If does arrive then I will increment the count by the no of
+all its permutations ((A+B+C+...)!/(A!*B!*C!*...)
+where A is the no. of a's and B is the no. of b's and so on.
+For n=7 7-digit no. we can use beggars method to calculate the
+no. of cases we have to check which is 1C(16,9) which equals to 11440
+cases which is significatly less than checking 10 million numbers.
+Also for a n digit no. maximum sum of squares its digits can be 81*n.
+So if we will store the result of first n*100 no.'s we can use it
+every time.
+
+"""
+
+
+def squared_sum_of_digits(num: int) -> int:
+    """
+    returns the sum of square of digits of num
+    >>> squared_sum_of_digits(1)
+    1
+    >>> squared_sum_of_digits(77)
+    98
+    >>> squared_sum_of_digits(11067)
+    87
+    """
+    sum = 0
+    while num:
+        sum += (num % 10) * (num % 10)
+        num //= 10
+
+    return sum
+
+
+def arrived_at_89(num: int) -> bool:
+    """
+    returns whether num will arrive at 89 or not
+    >>> arrived_at_89(10)
+    False
+    >>> arrived_at_89(19872)
+    True
+    >>> arrived_at_89(55555)
+    True
+
+    """
+    if num == 0:
+        return False
+
+    while num != 89 and num != 1:
+        num = squared_sum_of_digits(num)
+
+    if num == 89:
+        return True
+    else:
+        return False
+
+
+def all_possible_distributions(
+    all_distributions: list, each_distribution: list, current: int, index: int
+) -> None:
+    """
+    Appends the list of all the possible ways in which an n digts can
+    be distributed among 9 digits at the end of a 2-d list all_distributions.
+    >>> all_possible_distributions([],[0]*10,5,0) is None
+    True
+    >>> all_possible_distributions([],[0]*10,8,0) is None
+    True
+
```

Comment:
Please provide a better description regarding what the function is doing.

Also, the doctests are redundant. In this function, you're mutating the parameter which is not a good practice. Instead, what you can do is to make a copy of the list and mutate that list and return it. This way you will be able to test this function appropriately.

Reply:
The function appends the total no, of ways in which the number _**current**_ can be distributed among the indices 0-9 of the list. i.e when we add all the contents of each_distribution it will lead to **_current._** The each_distribution list is just a temporary list used to create a particular distribution.The all_distributions is a list of all possible current_distributions.  The index starts from 0 and the function recursively increases the index by 1 and terminates when it reaches 9 and then appends the formed each_distribution list to all_distributions list.

Reply:
Copying the 2-d list at every recursive call will increase the complexity by a large factor. Though it will be able to solve the problem within the given constraints it will not be able to generalize for a large input taking a toll on the performance of the algorithm.

Reply:
Also the function can be modified to return a large 2-d list of all possible distributions but in that case reformatting the returned 2-d list to so that there would max 88 characters in each line cause doctest to fail.
![ss](https://user-images.githubusercontent.com/39304055/103082512-46ff1a80-4600-11eb-96b5-7c88fbd6e0c7.png)

Reply:
https://docs.python.org/3/library/doctest.html#doctest.NORMALIZE_WHITESPACE

Reply:
> Copying the 2-d list at every recursive call will increase the complexity by a large factor. Though it will be able to solve the problem within the given constraints it will not be able to generalize for a large input taking a toll on the performance of the algorithm.

The function uses brute force method to generate all possible combinations ,thus it would make the function very slow. Should I still consider doing it ?

Reply:
Maybe you can copy the list first, mutate and finally append to the result for this function? @sharmapulkit04

Reply:
I am unable to understand what you are saying @poyea . Are you suggesting to copy the 2-d list at start of function call? If that is the case then what I am trying to say is that the function uses recursion to calculate all possible configurations, so if I copy the entire 2-d list at every recursive call, it will increase the complexity by a large factor and will make the function very slow.